### PR TITLE
Improvement of BasicStatistics

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
see #731 

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

File: framework/DataObjects/DataSet.py [hunk 1]
```diff
@@ -631,21 +633,25 @@ def write(self,fileName,style='netCDF',**kwargs):
...
     else:
       self.raiseAnError(NotImplementedError,'Unrecognized write style: "{}"'.format(style))
     if not self.hierarchical and 'RAVEN_isEnding' in self.getVars():
       return len(self._data.where(self._data['RAVEN_isEnding']==True,drop=True)['RAVEN_isEnding'])
     else:
-      return len(self) # so that other entities can track which realization we've written
+      if self.sampleTag in self._data.sizes.keys():
+        return len(self) # so that other entities can track which realization we've written
+      else:
+        return 0 # not 'RAVEN_sample_ID' is found
```
Comment (PaulTalbot-INL) [inline framework/DataObjects/DataSet.py]: Did we ever discuss whether all dataobjects should have a sample ID?  I think it's a discussion worth having, and I really don't know the correct answer.

File: framework/DataObjects/DataSet.py [hunk 2]
```diff
@@ -657,6 +663,18 @@ def __len__(self):
...
     return self.size
 
+  @property
+  def isEmpty(self):
+    """
+      @ In, None
+      @ Out, boolean, True if the dataset is empty otherwise False
+    """
+    try:
+      empty = False if len(self.asDataset().variables) > 0 else True
```
Comment (PaulTalbot-INL) [inline framework/DataObjects/DataSet.py]: calling asDataset() is not particularly cheap.  Preference should be given to checking the size of the collector and data.

File: framework/DataObjects/DataSet.py [hunk 3]
```diff
@@ -1273,6 +1291,42 @@ def _fromNetCDF(self,fileName, **kwargs):
     for key,val in self._data.attrs.items():
       self._meta[key] = pk.loads(val.encode('utf-8'))
 
+  def _fromXarrayDataset(self,dataset):
+    """
+    """
+    if self._collector is not None or self._data is not None:
```
Comment (PaulTalbot-INL) [inline framework/DataObjects/DataSet.py]: Is this the same as "isEmpty"?

File: doc/user_manual/postprocessor.tex [hunk 4]
```diff
@@ -149,6 +149,19 @@ \subsubsection{BasicStatistics}
   quantities are going to be calculated, if \textit{False} unbiased.
   \default{False}
   %
+  \item \xmlNode{dataset}, \xmlDesc{string (boolean), optional field}, if \textit{True} \xmlString{DataSet}
```
Comment (alfoa) [inline doc/user_manual/postprocessor.tex]: is boolean.
The user should not care how we read it :D

File: doc/user_manual/postprocessor.tex [hunk 5]
```diff
@@ -149,6 +149,19 @@ \subsubsection{BasicStatistics}
...
+    will be used to store the calculation results, if \textit{False} \xmlString{PointSet} or \xmlString{HistorySet}
+    will be used to store the calculation results.
+    \nb The optional \xmlString{DataSet} is added only to this Post-Processor, one can still use the \xmlString{OutStreams}
```
Comment (alfoa) [inline doc/user_manual/postprocessor.tex]: ```Post-Processor``` -> ```PostProcessor``` (for consistency regarding the naming convention)

File: doc/user_manual/postprocessor.tex [hunk 6]
```diff
@@ -149,6 +149,19 @@ \subsubsection{BasicStatistics}
...
+    to print the variables available in the \textit{DataSet}. The \xmlString{"metric"} names are used as the
+    variable names, i.e. variable names listed in \xmlNode{Input} or \xmlNode{Output}. If scalar metrics, such as 
+    \xmlNode{expectedValue} and \xmlNode{variance}, are requested, the index variable \xmlString{targets} will be required.
+    If vector metrics, such as \xmlNode{covariance} and \xmlNode{sensitivity}, are requested, the index variables
+    \xmlString{targets} and \xmlString{features} will be required. If \xmlNode{percentile} is requested, an additional
+    index variable \xmlString{percent} should be added. In addition, if dynamic BasicStatistics (e.g. time) is requested,
+    the index variable \xmlString{time}  will be required.
```
Comment (alfoa) [inline doc/user_manual/postprocessor.tex]: you should put an example XML section...It is quite hard to follow. Maybe a bulleted list might work better.
Eg.

- If scalar metrics, such as  \xmlNode{expectedValue} and \xmlNode{variance} bla blas
- If percentile, bla bla
etc

File: framework/DataObjects/DataSet.py (hunk 2 above)
Comment (alfoa) [reply framework/DataObjects/DataSet.py]: agree

File: framework/PostProcessors/BasicStatistics.py [hunk 7]
```diff
@@ -143,6 +143,39 @@ def inputToInternal(self, currentInp):
...
+    # TODO: convert dict to dataset, I think this will be removed when DataSet is used by other entities that
+    # are currently using this Basic Statisitics PostProcessor.
+    if type(currentInput).__name__ == 'dict':
+      if 'targets' not in currentInput.keys():
+        self.raiseAnError(IOError, 'Did not find targets in the input dictionary')
+      inputDataset = xr.Dataset()
+      for var, val in currentInput['targets'].items():
+        inputDataset[var] = val
+      if 'metadata' in currentInput.keys():
+        metadata = currentInput['metadata']
```
Comment (alfoa) [inline framework/PostProcessors/BasicStatistics.py]: I thought that the ProbabilityWeight are not in the metadata anymore

File: tests/framework/user_guide/ravenTutorial/gold/stat/statisticalAnalysis_basicStatPP_time_dump_0.csv [hunk 8]
```diff
@@ -1,4 +1,5 @@
 time,skew_A,vc_A,percentile_5_A,percentile_95_A,mean_A,kurt_A,median_A,max_A,min_A,samp_A,var_A,sigma_A,nsen_A_sigma-A,nsen_A_decay-A,sen_A_sigma-A,sen_A_decay-A,pear_A_sigma-A,pear_A_decay-A,cov_A_sigma-A,cov_A_decay-A,vsen_A_sigma-A,vsen_A_decay-A
+0.0,-1.08297714942,2.2781295785e-16,1.0,1.0,1.0,-2.21764705882,1.0,1.0,1.0,20.0,5.18987437645e-32,2.2781295785e-16,2.40875840596e-32,4.85942523233e-53,0.0,0.0,7.10788996659e-17,8.73271251411e-17,4.9822794014e-30,5.80013659945e-40,5.26271872678e-35,9.63719664528e-46
```
Comment (alfoa) [inline tests/framework/user_guide/ravenTutorial/gold/stat/statisticalAnalysis_basicStatPP_time_dump_0.csv]: is this correct?

File: framework/DataObjects/DataSet.py (hunk 1 above)
Comment (wangcj05) [reply framework/DataObjects/DataSet.py]: The ```RAVEN_sample_ID``` is added.

File: framework/DataObjects/DataSet.py (hunk 2 above)
Comment (wangcj05) [reply framework/DataObjects/DataSet.py]: changed

File: framework/DataObjects/DataSet.py (hunk 3 above)
Comment (wangcj05) [reply framework/DataObjects/DataSet.py]: ```if not self.isEmpty``` is used here now

File: doc/user_manual/postprocessor.tex (hunk 4 above)
Comment (wangcj05) [reply doc/user_manual/postprocessor.tex]: changed

File: doc/user_manual/postprocessor.tex (hunk 5 above)
Comment (wangcj05) [reply doc/user_manual/postprocessor.tex]: changed

File: doc/user_manual/postprocessor.tex (hunk 6 above)
Comment (wangcj05) [reply doc/user_manual/postprocessor.tex]: updated

File: framework/PostProcessors/BasicStatistics.py (hunk 7 above)
Comment (wangcj05) [reply framework/PostProcessors/BasicStatistics.py]: This is for dictionary. In some other post-processor, dictionary is used as input to basic statistics postprocessor, and ```metadata``` are used to store probabilityweight. When all postprocessor are converted to use DataSet directly, this can be remove.

File: tests/framework/user_guide/ravenTutorial/gold/stat/statisticalAnalysis_basicStatPP_time_dump_0.csv (hunk 8 above)
Comment (wangcj05) [reply tests/framework/user_guide/ravenTutorial/gold/stat/statisticalAnalysis_basicStatPP_time_dump_0.csv]: For some reason, the zero time step is skipped in the devel branch. This is not analytic test, and the results look ok to me. By the way, all the analytics tests are green for this PR.

**Comment by alfoa** [comment]
----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code. ***ok***
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts). ***n/a. Just a new option has been added***
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details. ***ok***
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass. ***ok***
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True. ***test added***
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync. ***As Control Board's Chair I approve (modified BasicStatistics)***
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done. ***#731***
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added? ***n/a***

**Comment by alfoa** [comment]
Check list passed...

The machines TesOpenSUSE Leap 42, Test OpenSUSE Tumbleweed and Test CentOS 6 are still not functional so are considered "allowed to fail"

PR can be accepted and merged...
